dev-docs/RFCs/vNext/layer-extension-rfc.md

```diff
@@ -0,0 +1,151 @@
+
+To mitigate this, we can of course add as many customizable functionalities to the base layer as possible. However, it is a major concern that inflating the base layer will sacrefice performance in return. Additional attributes takes CPU cycles to generate and memory to store. Larger shaders take longer to compile. WebGL1 has a restricting limit of 16 attributes per vertex shader, so adding attributes without the user asking for it will make the layer less extensible instead.
+
+
+## Case Studies
+
+### fp64
+
+* Shader module: `project64` (replaces `project32`)
+
+### Data Filter
+
+See [Data Filter RFC](/dev-docs/RFCs/v6.0/data-filter-rfc.md)
+
+Components:
+
+* Props: `getFilterValue`, `filterRange`
+* Shader module: `filter`
+* Attributes: `instanceFilterValue`
+* Shader injection:
+  - Vertex shader: `filter_setVisibility(instanceFilterValue);`
+  - Frament shader: `gl_FragColor = filter_filterColor(gl_FragColor);`
+
+### Brushing
+
+See [ArcBrushingLayer example](https://gist.github.com/Pessimistress/dc2becf3809c67dc443b4dbab1b9a46f#file-index-html-L128)
+
+A generic version of this functionality would need the following components:
+
+* Props: `enableBrush`, `brushRadius`
+* Shader module: `brushing`
+* Shader injection:
+  - Vertex shader: `brush_setVisibility(instancePositions);`
+  - Frament shader: `gl_FragColor = brush_filterColor(gl_FragColor);`
+* Subscribe to hover events and store the mouse position in layer state
+
+## Proposal
+
+### Layer Class Changes
+
+Some changes to the base Layer class are needed to make a generic, reusable extension system work:
+
+* A base layer `getShaders` method that each layer can utilize to augment their shaders with extensions.
+* (Implemented in v7.1) Official `bufferLayout` support in attribute management. The `bufferLayout` state is used in PathLayer and PolygonLayer to describe the number of instance for each data object. Currently, `AttributeManager`'s auto-update feature always assumes 1:1 mapping between an instance and a data object. Without supporting variable layout, an extension will have to implement layer-specific updaters for these layers.
+
+### New LayerExtension Class
+
+Add a `LayerExtension` interface. All layer extensions should extend this class. It contains the following methods:
+
+- `getShaders(shaders)` - called after a layer's own `getShaders`, a hook to inject additional modules/code into the shaders
+- `initializeState(layer, initParams)` - called after a layer's own `initializeState`, a hook to add attributes and/or initial states
+- `updateState(layer, initParams)` - called after a layer's own `updateState`, a hook to update layer state from props
```

Comment:
I assume the `initParams` in `updateState` are the extension props? Maybe call them `extensionParameters`, `extensionProps` or even just `props`?

To somewhat streamline/generalize the extension mechanism, I would vote for not unnecessarily over-emphasize the connection to layers, and renaming these methods to
- `initialize(props, layer)`
- `update(props, layer)`
- `finalize(layer)`

Reply:
`initParams` are the same object that is passed to `layer.initializeState`. I will clarify.